numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -309,6 +317,85 @@ static int
 
 /**end repeat**/
 
+#ifdef HAVE_STRTOLD_L
+static NPY_INLINE npy_longdouble string_to_long_double(PyObject*op)
+{
+    char *s;
+    char *end;
+    npy_longdouble temp;
+    PyObject* b;
+
+    /* This weird structure is to make the error handling work:
+     * if it's not a string, we want to fall back to other
+     * approaches. In python 2, that happens when we try to
+     * convert the string to a C string. In python 3, we have
+     * to use a two-step conversion process, via bytes; the
+     * not-a-string situation gets triggered on conversion to 
+     * bytes, but we want to make sure bytes objects get 
+     * handled too, because python accepts bytes as input to
+     * float() (why?).
+     */
+#if PY_MAJOR_VERSION >= 3
```

Comment:
I have no experience with the python C API so I couldn't figure out how to write version-agnostic string code. But I think an ifdef is probably necessary to handle both byte and string objects in python 3.